app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .where(username_lower: names.map(&:downcase), staged: false)
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where('lower(name) IN (?)', names.map(&:downcase))
+      .index_by(&:name)
+
+    mentionable_group_ids = Group
+      .mentionable(current_user, include_public: false)
+      .where('lower(name) IN (?)', names.map(&:downcase))
+      .pluck(:id)
+      .to_set
+
+    members_visible_group_ids = Group
+      .members_visible_groups(current_user)
+      .where('lower(name) IN (?)', names.map(&:downcase))
+      .pluck(:id)
+      .to_set
+
+    topic_muted_by = if topic.present?
+      TopicUser
+        .where(topic: topic)
+        .where(user_id: users.values.map(&:id))
+        .where(notification_level: TopicUser.notification_levels[:muted])
+        .pluck(:user_id)
+        .to_set
+    else
+      Set.new
+    end
+
+    topic_allowed_user_ids = if allowed_names.present?
+      User
+        .where(username_lower: allowed_names.map(&:downcase))
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedUser
+        .where(topic: topic)
+        .pluck(:user_id)
+        .to_set
+    end
```

Comment:
Should there be an `else` branch which defaults to an empty set? 🤔 

```suggestion
    topic_allowed_user_ids = if allowed_names.present?
      User
        .where(username_lower: allowed_names.map(&:downcase))
        .pluck(:id)
        .to_set
    elsif topic&.private_message?
      TopicAllowedUser
        .where(topic: topic)
        .pluck(:user_id)
        .to_set
    else
      Set.new
    end
```

Reply:
`topic_allowed_user_ids` should be used only in the context of a PM. I think it is better to fail with an error than to fail silently.